### ISSUE TITLE
yarn workspace imported to IDE causes issues

### ISSUE THREAD

**Issue by yoosiba** [eclipse/n4js#984]
When importing projects with from git repository IDE is importing also top level project, in particular the yarn workspace. It is useful to do this, root location usually contains a lot of important scripts (e.g. all releng scripts) and top level dependencies. Having this in the IDE allows to edit and mange those from one place.

Unfortunately this breaks IDE behaviour in terms of validations, installing dependencies and other making IDE unusable. Breakage happens due to projects appearing as nested in other projects.

Note that this is N4JS IDE specific issue, as in JDT this does not happen. For example working in Eclipse it is possible to import top level `n4js` as well as nested `n4js/plugins/org.eclipse.n4js` and IDE still works as expected.

Supporting importing level folder, a.k.a. yarn workspace, would be great connivence for yarn based projects.

**Comment by mor-n4** [comment]
> Note that this is N4JS IDE specific issue, as in JDT this does not happen. For example working in Eclipse it is possible to import top level n4js as well as nested n4js/plugins/org.eclipse.n4js and IDE still works as expected.

There is a difference: the top-level "n4js" project is not a Java project, just a plain Eclipse project (with Maven nature, maybe). If having full-blown Java projects nested inside a higher-level Java project, the JDT might have issues too.

Still, it would be nice to support this for N4JS (and it seems almost mandatory once we have support for yarn workspaces). At the very least we should detect this and show an error message (current behavior is very confusing).

**Comment by yoosiba** [comment]
@mor-n4 yes you are right, project is different. Case above was worst mixture of both:
1.  top level project does not have `XtextNature` and it is not processed by the xtext builder
2. top level project has `package.json` 

Due to `1.` it is not real N4JS Project but plain Eclipse Project, but due to `2.` it is considered as N4JS Project by some internals of IDE (but not all of them, e.g. builder).

Note that adding proper nature to the top level to make it real N4JS project, just makes things worse.

Again, supporting importing level folder, a.k.a. yarn workspace, would be great connivence for yarn based projects.

**Comment by lbeurerkellner** [comment]
Some investigation and experimentation has lead me to the following results:

The overall malfunctioning that can be observed comes down to the fact that resource changes in the workspace, especially regarding resources that are contained by more than one workspace project (e.g. by a nested and by a top-level project), are not processed correctly. More specifically, changing `package.json` resources does not correctly invalidate the cache of `ProjectDescription` instances in the `EclipseBasedN4JSWorkspace`. This can lead to the very odd behavior, as observed by @yoosiba, that older versions of an external library are installed, even though a newer version is specified in the `package.json` files. So generally, this comes down to a cache invalidation issue.

The invalidation of said `ProjectDescription` cache, is usually triggered in the method `N4JSAllContainersState#isAffectingContainerState`. In this method, we expect Xtext to run all workspace-wide `IResourceDelta`s by our implementation, so we can check whether we need to invalidate the cached `ProjectDescription` for a specific project. In this method we filter for `package.json` files and invalidate the cache appropriately. However, the assumption that this method will be invoked for every occurring resource change, does not hold true. In fact, the moment Xtext receives the return value `true`, it will stop traversing the resource delta tree. This means, that we only invalidate the first discovered project and we may not always be called for the potential other changed `package.json` files.
